### PR TITLE
expose sdpa helpers to python

### PR DESCRIPTION
**Pull request by drisspg**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113103
* __->__ #110496

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### CODE REVIEW

File: aten/src/ATen/native/transformers/cuda/sdp_utils.cpp [hunk 1]
```diff
@@ -19,6 +19,7 @@
 #include <cmath>
 #include <functional>
 #include <iostream>
+#include "c10/macros/Export.h"
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:22]: fix

File: torch/csrc/cuda/Module.cpp [hunk 2]
```diff
@@ -871,6 +874,71 @@ PyObject* THCPModule_cudaGetSyncDebugMode(PyObject* self, PyObject* noargs) {
...
+      // const c10::optional<at::Tensor>, double, bool>())
+      .def(py::init([](py::object query,
+                       py::object key,
+                       py::object value,
+                       py::object attn_mask,
+                       double dropout,
+                       bool is_causal) {
+        // TODO: how do we handle these with more grace than seg fault
+        // TODO: (Other way ?)Check if pyobj attn_mask is none if not cast
+        if (attn_mask.ptr() == Py_None) {
```
Comment (drisspg) [inline torch/csrc/cuda/Module.cpp:895]: could be ternary for readability, but is there a better way

Comment (drisspg) [reply torch/csrc/cuda/Module.cpp:895]: Thanks jeffery

**Review by jbschlosser — COMMENTED** [review]
dumb Q: what's the reasoning behind not making these helpers ATen ops? naively, it seems like we'll need that when we introduce the attn bias abstraction via subclasses so we can dispatch properly

**Comment by drisspg** [comment]
The high level thinking, was it would be nice to be able to use the SDPAParams directly which torchgen doesn't support. As well these are only expected  to work with cuda devices. 

SO concurrently if the inputs are not cuda we will just realize the masks and call SDPA

**Comment by jbschlosser** [comment]
Gotcha thanks! Looks like these will not work for me if I pass NT subclass instances for `query, key, value`. One concrete example is `check_for_seq_len_0_and_consistent_head_dim_nested_tensor_helper`, which assumes C++ NT metadata.

I'm looking into implementing the backend selection logic from python for the NT subclass, and I'll need access to a lot of the checks for determining if e.g. flash is usable but it needs to be somewhat customizeable over what is done now.

**Comment by jbschlosser** [comment]
Discussed offline: we'll refactor checks for jagged in a follow-up PR

**Review by jbschlosser — APPROVED** [review]
stampy

File: torch/csrc/cuda/Module.cpp [hunk 3]
```diff
@@ -3,6 +3,7 @@
 #include <ATen/core/TensorBody.h>
 #include <ATen/cuda/CUDAConfig.h>
 #include <ATen/native/ConvUtils.h>
+#include <autograd/python_variable.h>
```
Comment (drisspg) [inline torch/csrc/cuda/Module.cpp:6]: remove

**Pull request closed without merging**